### ISSUE TITLE
How to prevent multipart/form-data uploading?

### ISSUE THREAD

**Issue by skotchio** [mailru/FileAPI#325]
I'm using `client_body_in_file_only` directive on my nginx that does't support `multipart/form-data` data. How could I switch to binary uploading instead `multipart/form-data` one?

I use the following js:

    onImageChange: function (e, data) {

       var files = FileAPI.getFiles(e.target);

       FileAPI.upload({
         uploadRetry: 3,
         autoUpload: true,
         url: "192.168.56.101:3001",
         files: { file: files[0] }
       });
    }
        
The request headers are the following:

    Accept:*/*
    Accept-Encoding:gzip, deflate
    Cache-Control:no-cache
    Connection:keep-alive
    Content-Length:131394
    Content-Type:multipart/form-data; boundary=----WebKitFormBoundaryk2OCuEdfGrnnXP0W
    Host:192.168.56.101:3001
    Origin:http://192.168.56.101:3001
    Pragma:no-cache
    Referer:http://192.168.56.101:3001/upload
    X-Requested-With:XMLHttpRequest

**Comment by RubaXa** [comment]
This header sets the browser, rather than a library.

**Comment by skotchio** [comment]
Yes but with headers a library add `Request Payload`.Could you help me with my problem. How should I setup uploading on c client to send binary data?

**Comment by RubaXa** [comment]
No, it makes the browser, the library does not change `Content-Type`.
You can disable the use of FormData (http://mailru.github.io/FileAPI/#en/FileAPI.setup):
```js
var FileAPI = {
    formData: false
};
```


P.S. But this is a bad idea, better unuse nginx_module.

**Comment by skotchio** [comment]
Why this is bad idea I don't understand? And do you mean I have to use https://github.com/vkholodkov/nginx-upload-module instead?

**Comment by RubaXa** [comment]
Because it does only supports `multipart/form-data`.

**Comment by skotchio** [comment]
I've tried your approach and data saved to file via `client_body_in_file_only` looks like the following: http://screencast.com/t/fg7erxpNdpo How to strip the following info ?

```
--_fileapi1428823315567
Content-Disposition: form-data; name="file"; filename="47132788_138.jpg"
Content-Type: image/jpeg
```

**Comment by RubaXa** [comment]
And what you do not like? This is the correct `multipart/form-data`:

https://github.com/mailru/FileAPI/blob/master/lib/FileAPI.XHR.js#L362-L374

P.S Wow! This nginx module has not been updated 7 years! Find something modern or use any opportune server-side language for processing uploading.

**Comment by skotchio** [comment]
> And what you do not like? This is the correct 
Are you sure this is correct? These data should be sent via headers instead body file because these don't relate to image itself and corrupt it.

**Comment by RubaXa** [comment]
Absolutely sure.

**Comment by skotchio** [comment]
Ok I'll try your approach in production and give you feedback. Only two questions:

1. Which browser support of `var FileAPI = { formData: false};` approach?
2. How do you handling uploading files in your company MailRu (nginx-upload-module, php level or something else)?

**Comment by RubaXa** [comment]
1. IE10+, Opera 12+, FireFox 4+, Chrome 10+, Safari 6+
2. In our company we use this library together with Perl, Java, PHP, Python and NodeJS (we have a lot of different projects)

**Comment by skotchio** [comment]
Thanks for the answers. I've found that `nginx-upload-module` has 2.2 branch that no so old https://github.com/vkholodkov/nginx-upload-module/tree/2.2

**Comment by skotchio** [comment]
I've just tried to upload and get back jpg image and it seems corrupt. Please see the following link http://www.filedropper.com/0000000218 where I've just uploaded image uploading via `var FileAPI = { formData: false};` approach. If you try to open up this image that you see it's corrupt.

**Comment by RubaXa** [comment]
Probably, file not properly processed upon receipt.

**Comment by skotchio** [comment]
No, I'm doing it write. I have the following nginx config:

    location = /upload {
       client_body_temp_path       /var/uploads;
       client_body_in_file_only    on;
       client_body_buffer_size     1024k;
       client_max_body_size        5M;

       proxy_pass_request_headers  on;
       proxy_set_header            X-FILE          $request_body_file;
       proxy_redirect              off;
       proxy_pass_request_body     off;
       proxy_set_header            Content-Length  0;
       proxy_http_version          1.1;
       proxy_pass                  http://upstream;
    }

`http://upstream` does't handle uploaded file but send `201 Created` response only.

**Comment by RubaXa** [comment]
I think you'd better give up the library FileAPI in favor of [jQuery File Upload](https://github.com/blueimp/jQuery-File-Upload/) (they officially support [nginx-upload-module](https://github.com/blueimp/jQuery-File-Upload/wiki/Uploading-to-nginx-using-the-nginx-upload-module)).

**Comment by skotchio** [comment]
Yes I know. But I like your library and want to use in production. Did you open the image that I've uploaded to http://www.filedropper.com/0000000218 ?

**Comment by RubaXa** [comment]
I honestly do not know what the reason, but the library is covered by tests,  `formData: false` checked through PhantomJS.

**Comment by skotchio** [comment]
So you can open uploaded image thought `formData: false` ?

**Comment by RubaXa** [comment]
Yes, the image broken, but I think that this is a problem `nginx-upload-module`.

**Comment by skotchio** [comment]
I don't use `nginx-upload-module` instead I use `client_body_in_file_only` nginx directive.

**Comment by RubaXa** [comment]
My knowledge of `nginx` end on `client_max_body_size`.

**Comment by skotchio** [comment]
@RubaXa I want to solve the problem and share experience to all. Do you have someone who could understand what is a problem?  If we get it work it safe a lot of time to other people and make your module more popular

**Comment by RubaXa** [comment]
Alas, this open source project, I can only put a label "help wanted".

**Comment by LukeAbell** [comment]
I think we are experiencing this same problem. We're using hull.io, fileapi, and Squarespace.

**Comment by smtperece** [comment]
the problem with client_body_in_file_only is that this was not originally intended for file uploads. this is primarily debugging feature, which also can be used for 'layered' processing by several backends (until one recognizes that request as it's own) and other evaluation-grade, inefficient in runtime sense, but very efficient in development time solutions. it is NOT standard-conforming. neither RFC 2388, nor RFC 1867. nothing. it just saves request body as a whole, without any processing, into a file. file of Content-length bytes.
someone noticed that with specially tailored client this can be used for upload handling.
and since file bodies are processed with frontend (nginx itself) this saves extra trip from frontend to backend for all these data. and this is "architecturally good" solution.
but it REQUIRES nonconformant, custom client.
library in question, on contrary, tries to mimic browser behavior, then browser submits form with enctype="multipart/form-data". it is an RFC2388-compliant client. these two just can't play nice with each other.

nginx-upload-module is a standard-compliant server-side module. but there is another catch.
nginx-upload-module is not in nginx-core. It is not part of nginx distribution. it is "contribware".
to my best knowledge, it has a separate maintainer, and not actively maintained at all these days.
it relies on some internals, changing from time to time. so even if You'll find a working version for Your particular version of nginx (which is not always an easy task, either) it can stop working right after next upgrade.
and with that in mind (and only in that sense) people calling client_body_in_file_only "more robust" and "more production ready".
if You need to parse a POST generated by browser itself, or if Your request comes from standard-compliant library, You need either nginx-upload-module or just handle uploads in backend...

and, as a side-note: since nginx-upload-module is standard-compliant, it does not need any "official support" from library vendor.

hope this helps.
